src/charm.py

```diff
@@ -67,17 +70,28 @@ def _on_config_changed(self, event) -> None:
         self.configure_workload_container(event)
 
     def _report_visits_by_ip_action(self, event: ActionEvent) -> None:
-        """Report requests to nginx by IP and report action result."""
+        """Handle the report-visits-by-ip action."""
+        results = self._report_visits_by_ip(event)
+        event.set_results({"ips": str(results)})
+
+    def _report_visits_by_ip(self, _) -> list[(int, str)]:
+        """Report requests to nginx grouped and ordered by IP and report action result."""
         container = self.unit.get_container(CONTAINER_NAME)
-        process = container.exec(
-            [
-                "/bin/sh",
-                "-c",
-                "awk -vDate=`date -d'now-2 hours' +[%d/+%b/%Y:%H:%M:%S` '{ if ($4 > Date) print $1}' /var/log/nginx/access.log | sort  |uniq -c |sort -n | tail",  # noqa: E501
-            ]
+        log_path = "/var/log/nginx/access.log"
+        with container.pull(log_path) as log_file:
+            list_to_search = log_file.readlines()
+        ip_regex = r"[0-9]+(?:\.[0-9]+){3}"
+        ip_list = []
+        current_datetime = datetime.now()
+        start_datetime = current_datetime - timedelta(minutes=20)
+        for line in list_to_search:
+            line = line.split()
+            log_datetime = datetime.strptime(line[3].lstrip("[").rstrip("]"), "%d/%b/%Y:%H:%M:%S")
+            if log_datetime >= start_datetime and re.search(ip_regex, line[0]):
+                ip_list.append(str(line[0]))
+        return sorted(
```

Comment:
Ah I see, the intent is to return the number of hits of an IP address. There is a nice inbuilt data structure which can help with this: https://docs.python.org/3/library/collections.html#collections.Counter

It also means that the execution complexity is reduced since we shouldn't need to sort which usually has execution time complexity higher than O(N). A naive implementation would be to keep some sort of data structure around for each of the IP addresses and maintain a counter for how many times it has been seen, which has O(N) time complexity. The `Counter` data structure should help with that